modules/jfx.incubator.richtext/src/main/java/com/sun/jfx/incubator/scene/control/richtext/CellArrangement.java

```diff
@@ -0,0 +1,375 @@
+        }
+
+        return TextPos.ZERO;
+    }
+
+    /** returns the cell contained in this layout, or null */
+    public TextCell getCell(int modelIndex) {
+        int ix = modelIndex - origin.index();
+        if (ix < 0) {
+            if ((ix + topCount()) >= 0) {
+                // cells in the top part come after bottom part, and in reverse order
+                return cells.get(bottomCount - ix - 1);
+            }
+        } else if (ix < bottomCount) {
+            // cells in the normal (bottom) part
+            return cells.get(ix);
+        }
+        return null;
+    }
+
+    /** returns a visible cell, or null */
+    public TextCell getVisibleCell(int modelIndex) {
+        int ix = modelIndex - origin.index();
+        if ((ix >= 0) && (ix < visibleCount)) {
+            return cells.get(ix);
+        }
+        return null;
+    }
+
+    /** returns a TextCell from the visible or bottom margin parts, or null */
+    public TextCell getCellAt(int ix) {
+        if (ix < visibleCount) {
+            return cells.get(ix);
+        }
+        return null;
+    }
+
+    /**
+     * Creates a CaretInfo.
+     * @param target the target region (vflow.content)
+     * @param p the text position
+     * @return the new CaretInfo
+     */
+    public CaretInfo getCaretInfo(Region target, TextPos p) {
+        if (p != null) {
+            int ix = p.index();
+            TextCell cell = getCell(ix);
+            if (cell != null) {
+                int charIndex = p.charIndex();
+                boolean leading = p.isLeading();
+                PathElement[] path = cell.getCaretShape(target, charIndex, leading);
+                if (path != null) {
+                    double lineSpacing = cell.getLineSpacing();
+                    return CaretInfo.create(lineSpacing, path);
+                }
+            }
+        }
+        return null;
+    }
+
+    public void removeNodesFrom(Pane p) {
+        ObservableList<Node> cs = p.getChildren();
+        for (int i = getVisibleCellCount() - 1; i >= 0; --i) {
+            TextCell cell = cells.get(i);
+            cs.remove(cell);
+        }
+    }
+
+    public void setBottomCount(int ix) {
+        bottomCount = ix;
+    }
+
+    /** visible + bottom margin cells */
+    public int bottomCount() {
+        return bottomCount;
+    }
+
+    public int cellCount() {
+        return cells.size();
+    }
+
+    public void setBottomHeight(double h) {
+        bottomHeight = h;
+    }
+
+    /** in pixels from the first visible cell to the last cell in the arrangement */
+    public double bottomHeight() {
+        return bottomHeight;
+    }
+
+    public int topCount() {
+        return cells.size() - bottomCount;
+    }
+
+    public void setTopHeight(double h) {
+        topHeight = h;
+    }
+
+    public double topHeight() {
+        return topHeight;
+    }
+
+    public double averageHeight() {
+        int sz = cells.size();
+        if (sz == 0) {
+            return 20;
```

Comment:
Does this 20 need to scale on different uiScales?
Also - can we have this constant declared somewhere?